### PR TITLE
Allow dynamic publish permissions based on reply subjects of received msgs.

### PR DESCRIPTION
**Pull request by derekcollison**
This allows a dynamic publish permission to be granted only to subjects that were delivered as reply subjects. We also allow more detailed permission on maximum number of messages that can use the permission and maximum amount of time the permission can be active.

Signed-off-by: Derek Collison <derek@nats.io>

/cc @nats-io/core

### CODE REVIEW

**Comment by derekcollison** [comment]
I will add support at JWT level after this lands.

**Comment by derekcollison** [comment]
Also probably need a test for duplicate reply subjects and resetting of permissions. Also a more explicit test on the pruning.

File: server/client.go [hunk 1]
```diff
@@ -596,6 +608,13 @@ func (c *client) setPermissions(perms *Permissions) {
 		}
 	}
 
+	// Check if we are allowed to send responses.
+	if perms.Response != nil {
+		c.perms.resp = perms.Response
```
Comment (kozlovic) [inline server/client.go]: Is that safe? Should we make a copy/clone of perms.Response? I am thinking config reload, if permissions were to change. Maybe it's not relevant, but for other perms with insert in local sublist, here we store the reference..

File: server/client.go [hunk 2]
```diff
@@ -2131,6 +2150,15 @@ func (c *client) deliverMsg(sub *subscription, mh, msg []byte) bool {
 
 	client.out.pm++
 
+	// If we are tracking dynamic publish permissions that track reply subjects,
+	// do that accounting here. We only look at c.resps which will be non-nil.
```
Comment (kozlovic) [inline server/client.go]: wrong reference to `c.resps` in this comment?

File: server/client.go [hunk 3]
```diff
@@ -2205,11 +2254,31 @@ func (c *client) pubAllowed(subject string) bool {
...
-	c.perms.pcache[string(subject)] = allowed
-	// Prune if needed.
-	if len(c.perms.pcache) > maxPermCacheSize {
-		c.prunePubPermsCache()
+
+	// If we are currently not allowed but we are tracking reply subjects
+	// dynamically, check to see if we are allowed here Avoid pcache.
+	// We need to acquire the lock though.
+	if !allowed && c.perms.resp != nil {
+		c.mu.Lock()
```
Comment (kozlovic) [inline server/client.go]: That is going to be a problem. `pubAllowed` is called by `canImport` that is holding the connection's lock. Let's check that.

File: server/opts.go [hunk 4]
```diff
@@ -1980,6 +2003,52 @@ func parseSubjects(v interface{}, errors, warnings *[]error) ([]string, error) {
 	return subjects, nil
 }
 
+// Helper function to parse new style authorization into a SubjectPermission with Allow and Deny.
```
Comment (kozlovic) [inline server/opts.go]: Look like wrong comment...

File: server/client.go (hunk 1 above)
Comment (derekcollison) [reply server/client.go]: We could make a copy, we do not have sublists it just 2 fields, and int and a time.Duration.

File: server/client.go (hunk 2 above)
Comment (derekcollison) [reply server/client.go]: Will fix.

File: server/client.go (hunk 3 above)
Comment (derekcollison) [reply server/client.go]: ok will take a closer look.

File: server/opts.go (hunk 4 above)
Comment (derekcollison) [reply server/opts.go]: will fix.

File: server/client.go (hunk 1 above)
Comment (derekcollison) [reply server/client.go]: ok fixed. Will squash down all the commits when we sign off.

File: server/client.go (hunk 3 above)
Comment (kozlovic) [reply server/client.go]: But then this means that when invoked from a route, that won't work. That is, if we allow these kind of permissions for routes, then they won't be checked.

Comment (kozlovic) [reply server/client.go]: instead, I would have conditionally lock/unlock since in one case we already have the lock, other we don't.

Comment (derekcollison) [reply server/client.go]: I disallowed from cluster config.
